pkg/download/consistent_hashing.go

```diff
@@ -285,23 +307,68 @@ func (m *ConsistentHashingMode) consistentHashIfNeeded(req *http.Request, start
 			// later without breaking things.
 			// note that it's not safe to share a HashOptions so we create a fresh one each time.
 			hashopts := &hashstructure.HashOptions{IgnoreZeroValue: true}
-			hash, err := hashstructure.Hash(key, hashstructure.FormatV2, hashopts)
+			var hash uint64
+			hash, err = hashstructure.Hash(key, hashstructure.FormatV2, hashopts)
 			if err != nil {
-				return fmt.Errorf("error calculating hash of key")
+				err = fmt.Errorf("error calculating hash of key: %w", err)
+				return
 			}
 
 			logger.Debug().Uint64("hash_sum", hash).Int("len_cache_hosts", len(m.CacheHosts)).Msg("consistent hashing")
 
 			// jump is an implementation of Google's Jump Consistent Hash.
 			//
 			// See http://arxiv.org/abs/1406.2294 for details.
-			cachePodIndex := int(jump.Hash(hash, len(m.CacheHosts)))
+			cachePodIndex = int(jump.Hash(hash, len(m.CacheHosts)))
 			cacheHost := m.CacheHosts[cachePodIndex]
 			logger.Debug().Str("cache_key", fmt.Sprintf("%+v", key)).Int64("start", start).Int64("end", end).Int64("slice_size", m.SliceSize).Int("bucket", cachePodIndex).Msg("consistent hashing")
 			if cacheHost != "" {
 				req.URL.Scheme = "http"
 				req.URL.Host = cacheHost
 			}
+			return
+		}
+	}
+	return
+}
+
+func (m *ConsistentHashingMode) retryConsistentHash(req *http.Request, start int64, end int64, originalIndex int) error {
+	if len(m.CacheHosts) == 1 {
+		return fmt.Errorf("Can't retry with only one cache host")
+	}
+	logger := logging.GetLogger()
+	for _, host := range m.DomainsToCache {
+		if host == req.URL.Host {
+			if start/m.SliceSize != end/m.SliceSize {
+				return fmt.Errorf("can't make a range request across a slice boundary: %d-%d straddles a slice boundary (slice size is %d)", start, end, m.SliceSize)
+			}
+			slice := start / m.SliceSize
+
+			key := CacheKey{URL: req.URL, Slice: slice, Retry: true}
+			// we set IgnoreZeroValue so that we can add fields to the hash key
+			// later without breaking things.
+			// note that it's not safe to share a HashOptions so we create a fresh one each time.
+			hashopts := &hashstructure.HashOptions{IgnoreZeroValue: true}
+			hash, err := hashstructure.Hash(key, hashstructure.FormatV2, hashopts)
+			if err != nil {
+				return fmt.Errorf("error calculating hash of key: %w", err)
+			}
+
+			logger.Debug().Uint64("hash_sum", hash).Int("len_cache_hosts", len(m.CacheHosts)).Msg("consistent hashing")
+
+			// jump is an implementation of Google's Jump Consistent Hash.
+			//
+			// See http://arxiv.org/abs/1406.2294 for details.
+
+			// we advance around the ring by somewhere between 1 and n-1
+			displacement := int(jump.Hash(hash, len(m.CacheHosts)-1)) + 1
```

Comment:
As with my comment above, this isn't necessarily a blocker right now but I'm a little bit uneasy that we have the actual consistent hash implementation duplicated in multiple places. It feels this (and the hashstructure/key calculations) should probably only be defined once and then if necessary used by glue code that uses the HTTP client.

Reply:
done